### PR TITLE
feat: add nano contract register screen [13]

### PR DESCRIPTION
**Pull request by alexruzenhack**
### Acceptance Criteria
- it should add `NanoContractRegisterScreen`
- it should add action `NANOCONTRACT_REGISTER_READY`
- it should improve `FeedbackModal` component
- it should improve `HathorModal` component
- it should refactor the transmission of `ncId` to `NanoContractDetailsScreen`
- it should add reactions to register actions on reducer
- it should add some small fixes and tweaks
- It should add feedback content on first address loading

Nano Contract Registration

https://github.com/HathorNetwork/hathor-wallet-mobile/assets/5992210/87f8308b-b1e2-4759-8805-759568e827d5

First Address Loading

First Address Error

>[!NOTE]
>It depends on:
>- https://github.com/HathorNetwork/hathor-wallet-mobile/pull/440

### Security Checklist
- [x] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

### CODE REVIEW

File: src/components/NanoContract/NanoContractDetails.component.js [hunk 1]
```diff
@@ -43,7 +43,7 @@ const getNanoContractDetails = (ncId) => (state) => {
    * quickly, therefore the user will not have time to see the default state.
```
Comment (andreabadesso) [inline src/components/NanoContract/NanoContractDetails.component.js:43]: Are you going to rename the component (to remove `.component.` in this PR?

File: src/components/NanoContract/NanoContractDetails.component.js [hunk 2]
```diff
@@ -43,7 +43,7 @@ const getNanoContractDetails = (ncId) => (state) => {
...
    */
   const defaultMeta = { isLoading: false, error: null, after: null };
-  const txHistory = Object.values(state.nanoContract.history[ncId] || {});
+  const txHistory = Object.values(state.nanoContract.history[ncId] || []);
```
Comment (andreabadesso) [inline src/components/NanoContract/NanoContractDetails.component.js:46]: I think we discussed it last week, but you were going to remove this `Object.values`, right?

File: src/screens/NanoContract/helper.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+/**
+ * Check if the nano contract register status is failed.
+ * @param {object} status - status from redux store
+ * @returns {boolean} - true if the status is failed, false otherwise
+ */
+// eslint-disable-next-line max-len
+export const hasFailed = (status) => status === NANOCONTRACT_REGISTER_STATUS.FAILED;
+
```
[the remark this answers is not in the document]
Comment (andreabadesso) [reply src/screens/NanoContract/helper.js:29]: No need to ignore the max-len directive to save 1 line

**Review by andreabadesso — CHANGES_REQUESTED** [review]

File: src/components/NanoContract/NanoContractDetails.component.js (hunk 2 above)
Comment (alexruzenhack) [reply src/components/NanoContract/NanoContractDetails.component.js:46]: Yes, I removed it.

File: src/sagas/nanoContract.js [hunk 4]
```diff
@@ -75,6 +75,9 @@ export function* init() {
 export function* registerNanoContract({ payload }) {
   const { address, ncId } = payload;
 
+  // This delay protects this process against user jitter
+  yield delay(720);
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/sagas/nanoContract.js:79]: I introduced it because it runs quickly and in case a Nano Contract is already registered one can retry retry many times quickly. But maybe it is something unnecessary. I'm removing it.

File: src/screens/NanoContract/helper.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import { NANOCONTRACT_REGISTER_STATUS } from '../../constants';
+
+export const feedbackSucceedText = t`Contract successfully registered.`;
+
+/**
+ * Check if the nano contract register status is successful.
+ * @param {object} status - status from redux store
+ * @returns {boolean} - true if the status is successful, false otherwise
+ */
+export function hasSucceed(status) {
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/screens/NanoContract/helper.js:18]: Fixed.

File: src/screens/NanoContract/helper.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/screens/NanoContract/helper.js:29]: Indeed. Fixed.

File: src/screens/NanoContract/NanoContractRegisterScreen.js [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+    dispatch(nanoContractRegisterReady());
+    navigation.replace('NanoContractDetailsScreen', { ncId: formModel.ncId });
+  }
+
+  useEffect(() => {
+    const fetchData = async () => {
+      const defaultAddress = await wallet.getAddressAtIndex(0);
+      setAddress(defaultAddress);
+    };
+    fetchData();
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/screens/NanoContract/NanoContractRegisterScreen.js]: Refactored. I moved it to saga.

File: src/screens/NanoContract/NanoContractRegisterScreen.js [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+    ncId: null,
+  });
+  const [invalidModel, setInvalidModel] = useState({
+    ncId: null,
+  });
+
+  /* @param {
+   *   'nodeUrl'|'explorerUrl'|'explorerServiceUrl'|'walletServiceUrl'|'walletServiceWsUrl'
+   * } name */
+  const handleInputChange = (name) => (value) => {
```
Comment (andreabadesso) [inline src/screens/NanoContract/NanoContractRegisterScreen.js:76]: Since this method is being passed down to components as a prop, you should use the `useCallback` hook to memoize it, example:

```
  const handleInputChange = useCallback(
    (name) => (value) => {
      if (isClean) {
        setClean(false);
      }

      // update invalid model
      const invalidModelCopy = { ...invalidModel };
      delete invalidModelCopy[name];
      setInvalidModel(invalidModelCopy);

      // update form model
      const form = {
        ...formModel,
        [name]: value,
      };
      setFormModel(form);

      // validate form model and update invalid model
      setInvalidModel(validate(form));
    },
    [isClean, invalidModel, formModel]
  );
```

this will prevent unnecessary re-renders on the child component when the NanoContractRegisterScreen is re-rendered but the dependencies don't change (isClean, invalidModel, formModel)

File: src/screens/NanoContract/NanoContractRegisterScreen.js [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+      ...formModel,
+      [name]: value,
+    };
+    setFormModel(form);
+
+    // validate form model and update invalid model
+    setInvalidModel(validate(form));
+  };
+
+  const handleSubmit = () => {
```
Comment (andreabadesso) [inline src/screens/NanoContract/NanoContractRegisterScreen.js:97]: Same here

**Review by andreabadesso — CHANGES_REQUESTED** [review]

File: src/sagas/mixins.js [hunk 9]
```diff
@@ -49,8 +51,34 @@ export function* fetchAllWalletAddresses() {
...
 
+export function* fetchFirstWalletAddress() {
+  const wallet = yield select((state) => state.wallet);
+  if (!wallet.isReady()) {
+    log.error('Fail fetching first wallet address because wallet is not ready yet.');
+    const errorMsg = t`Wallet is not ready to load the first address.`;
+    // This will show the message in the feedback content
+    yield put(firstAddressFailure({ error: errorMsg }));
+    // This will show user an error modal with the option to send the error to sentry.
+    yield put(onExceptionCaptured(new Error(errorMsg), false));
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/sagas/mixins.js:62]: Oh! Good point.

[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/sagas/mixins.js:62]: Done.

File: src/screens/NanoContract/NanoContractRegisterScreen.js [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+      setInvalidModel(newInvalidModel);
+      return;
+    }
+
+    const { ncId } = formModel;
+    dispatch(nanoContractRegisterRequest({ address, ncId }));
+  };
+
+  const handleFeedbackModalDismiss = () => {
+    dispatch(nanoContractRegisterReady());
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/screens/NanoContract/NanoContractRegisterScreen.js:109]: Yes and no. The ready action sets the screen status on redux by updating `registerStatus`, and it impacts on components rendering on screen. The ready state shows the form ready to be fulfilled.

File: src/screens/NanoContract/NanoContractRegisterScreen.js (hunk 7 above)
Comment (alexruzenhack) [reply src/screens/NanoContract/NanoContractRegisterScreen.js:76]: Good point. Thank you!

Comment (alexruzenhack) [reply src/screens/NanoContract/NanoContractRegisterScreen.js:76]: Done.

File: src/components/NanoContract/NanoContractDetails.component.js (hunk 1 above)
Comment (alexruzenhack) [reply src/components/NanoContract/NanoContractDetails.component.js]: No. I can do it in another PR.

Comment (alexruzenhack) [reply src/components/NanoContract/NanoContractDetails.component.js]: https://github.com/HathorNetwork/hathor-wallet-mobile/pull/498

File: src/reducers/reducer.js [hunk 11]
```diff
@@ -7,7 +7,7 @@
 
 import hathorLib from '@hathor/wallet-lib';
 import { get } from 'lodash';
-import { INITIAL_TOKENS, DEFAULT_TOKEN, PUSH_API_STATUS, FEATURE_TOGGLE_DEFAULTS, PRE_SETTINGS_MAINNET, NETWORKSETTINGS_STATUS } from '../constants';
+import { INITIAL_TOKENS, DEFAULT_TOKEN, PUSH_API_STATUS, FEATURE_TOGGLE_DEFAULTS, PRE_SETTINGS_MAINNET, NETWORKSETTINGS_STATUS, NANOCONTRACT_REGISTER_STATUS } from '../constants';
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/reducers/reducer.js]: Done.

File: locale/pt-br/texts.po [hunk 12]
```diff
@@ -970,6 +971,82 @@ msgstr "A wallet não está pronta para carregar endereços."
...
+
+#: src/screens/NanoContract/NanoContractTransactionScreen.js:37
+msgid "Nano Contract Transaction"
+msgstr "Transação do Nano Contract"
+
+#: src/screens/NanoContract/helper.js:11
+msgid "Contract successfully registered."
+msgstr "Nano Contract registrado com sucesso."
+
+#: src/sagas/mixins.js:33
```
Comment (andreabadesso) [inline locale/pt-br/texts.po:1032]: This file is gone now, is an update missing on the translations?

Comment (andreabadesso) [reply locale/pt-br/texts.po:1032]: Same for all below

File: src/screens/NanoContract/NanoContractRegisterScreen.js [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+import Spinner from '../../components/Spinner';
+import FeedbackModal from '../../components/FeedbackModal';
+import errorIcon from '../../assets/images/icErrorBig.png';
+import checkIcon from '../../assets/images/icCheckBig.png';
+import { FeedbackContent } from '../../components/FeedbackContent';
+
+/**
+ * Verifies if the invalidModel of the form has an error message.
+ */
+function hasError(invalidModel) {
```
Comment (andreabadesso) [inline src/screens/NanoContract/NanoContractRegisterScreen.js:53]: Should this also be a helper?

**Review by andreabadesso — CHANGES_REQUESTED** [review]

File: locale/pt-br/texts.po [hunk 14]
```diff
@@ -970,6 +971,82 @@ msgstr "A wallet não está pronta para carregar endereços."
...
+msgid "Nano Contract ID is required."
+msgstr "ID do Nano Contract é obrigatório."
+
+#: src/screens/NanoContract/NanoContractRegisterScreen.js:135
+msgid "See contract"
+msgstr "Ver contrato"
+
+#: src/screens/NanoContract/NanoContractRegisterScreen.js:153
+msgid "Load First Addresses Error"
+msgstr ""
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply locale/pt-br/texts.po]: Translated.

File: locale/pt-br/texts.po (hunk 12 above)
Comment (alexruzenhack) [reply locale/pt-br/texts.po]: Removed.

File: src/screens/NanoContract/NanoContractRegisterScreen.js (hunk 13 above)
Comment (alexruzenhack) [reply src/screens/NanoContract/NanoContractRegisterScreen.js]: Moved to utils.

File: src/reducers/reducer.js [hunk 15]
```diff
@@ -323,6 +333,16 @@ const initialState = {
     addresses: [],
     error: null,
   },
+  /**
+   * firstAddress {{
+   *   address?: string;
+   *   error?: string;
```
Comment (andreabadesso) [inline src/reducers/reducer.js:339]:
```suggestion
   *   address: string | null;
   *   error: string | null;
```

If these are `undefined` (as implied by the `?`) the components will glitch, right?

File: src/reducers/reducer.js [hunk 16]
```diff
@@ -1326,28 +1358,76 @@ export const onNetworkSettingsUpdateInvalid = (state, { payload }) => ({
...
+    registerFailureMessage: error,
+  },
+});
+
 /**
  * @param {Object} state
  * @param {{
  *   payload: {
- *     entryKey: string,
+ *     entryKey: string;
```
Comment (andreabadesso) [inline src/reducers/reducer.js:1389]: Should you also change the attributes of `entryValue` as well?

**Review by andreabadesso — CHANGES_REQUESTED** [review]

File: locale/da/texts.po [hunk 17]
```diff
@@ -1305,20 +1357,24 @@ msgstr ""
 msgid "Unregister contract"
 msgstr ""
 
-#: src/components/NanoContract/NanoContractTransactionHeader.component.js:93
+#.  XXX: add <ArrowUpIcon /> when shrank component can be used.
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply locale/da/texts.po:1360]: Yes, it was added automatically. The first time I removed it, but it has appeared again and I left it there.

File: locale/pt-br/texts.po [hunk 18]
```diff
@@ -1041,6 +1087,26 @@ msgstr "Transação"
...
+
+#: src/sagas/wallet.js:770
+msgid "Wallet is not ready to load the first address."
+msgstr "A wallet não está pronta para carregar primeiro endereço."
+
+#.  This will show the message in the feedback content
+#: src/sagas/wallet.js:786
+msgid "There was an error while loading first wallet address. Try again."
+msgstr ""
+"Ocorreu um erro durante download do primeiro endereço da wallet. Tente novamente."
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply locale/pt-br/texts.po:1108]: Done.

File: locale/pt-br/texts.po [hunk 19]
```diff
@@ -1041,6 +1087,26 @@ msgstr "Transação"
...
 
+#: src/sagas/wallet.js:744
+msgid "Wallet is not ready to load addresses."
+msgstr "A wallet não está pronta para carregar endereços."
+
+#.  This will show the message in the feedback content at SelectAddressModal
+#: src/sagas/wallet.js:760
+msgid "There was an error while loading wallet addresses. Try again."
+msgstr ""
+"Ocorreu um erro durante download dos endereços da wallet. Tente novamente."
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply locale/pt-br/texts.po:1098]: Done.

File: locale/pt-br/texts.po [hunk 20]
```diff
@@ -1041,6 +1087,26 @@ msgstr "Transação"
...
+
+#: src/sagas/wallet.js:770
+msgid "Wallet is not ready to load the first address."
+msgstr "A wallet não está pronta para carregar primeiro endereço."
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply locale/pt-br/texts.po:1102]: Done.

File: src/reducers/reducer.js (hunk 15 above)
Comment (alexruzenhack) [reply src/reducers/reducer.js:339]: Why not use the `?` notation to indicate a nullable property? We use this notation in other docstrings as well.

When `address` is null and `error` is also null, then the component shows a loading feedback content. It should never glitch.

File: src/components/HathorModal.js [hunk 21]
```diff
@@ -38,9 +38,9 @@ const styles = StyleSheet.create({
     borderRadius: 8,
     alignItems: 'center',
     paddingHorizontal: 16,
-    paddingBottom: 56,
```
Comment (pedroferreira1) [inline src/components/HathorModal.js:41]: Why did you change this? You changed to less than half the original value, what screens are impacted with this?

File: src/sagas/nanoContract.js [hunk 22]
```diff
@@ -117,14 +116,13 @@ export function* registerNanoContract({ payload }) {
   const nc = {
     address,
     ncId,
-    blueprintId: ncState.blueprint_id,
```
Comment (pedroferreira1) [inline src/sagas/nanoContract.js:120]: Why did you remove this? The method `wallet.storage.registerNanoContract` expects a `INcData` as parameter, which has the `bluerprintId` attribute.

File: src/screens/NanoContract/helper.js [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+ */
+
+import { t } from 'ttag';
+import { NANOCONTRACT_REGISTER_STATUS } from '../../constants';
+
+export const feedbackSucceedText = t`Contract successfully registered.`;
+
+/**
+ * Check if the nano contract register status is successful.
+ * @param {object} status - status from redux store
```
Comment (pedroferreira1) [inline src/screens/NanoContract/helper.js:15]: Is the status an object, or a string? This is valid for all these methods.

File: src/screens/NanoContract/NanoContractRegisterScreen.js [hunk 24]
```diff
@@ -0,0 +1,323 @@
...
+              <CircleInfoIcon size={20} color='hsla(203, 100%, 25%, 1)' />
+            </View>
+            <View style={styles.infoContent}>
+              <Text style={styles.text}>
+                {t`If you want to change the wallet address, you will be able to do`}
+                <Text style={styles.bold}>
+                  {' '}{t`after the contract is registered.`}
+                </Text>
+              </Text>
+            </View>
```
Comment (pedroferreira1) [inline src/screens/NanoContract/NanoContractRegisterScreen.js:207]: I think this view should have a padding right as well (at least 4). The text doesn't seem to fit well, because it's really close to the border. 

image

File: src/screens/NanoContract/NanoContractRegisterScreen.js [hunk 25]
```diff
@@ -0,0 +1,323 @@
...
+  const [isClean, setClean] = useState(true);
+  const [formModel, setFormModel] = useState({
+    ncId: null,
+  });
+  const [invalidModel, setInvalidModel] = useState({
+    ncId: null,
+  });
+
+  /* @param {
+   *   'nodeUrl'|'explorerUrl'|'explorerServiceUrl'|'walletServiceUrl'|'walletServiceWsUrl'
```
Comment (pedroferreira1) [inline src/screens/NanoContract/NanoContractRegisterScreen.js:81]: What's this?

File: src/screens/NanoContract/NanoContractRegisterScreen.js [hunk 26]
```diff
@@ -0,0 +1,323 @@
...
+      <NavigationHeader navigation={navigation} />
+
+      {hasSucceeded(registerState.registerStatus)
+        && (
+          <FeedbackModal
+            icon={(<Image source={checkIcon} style={styles.feedbackModalIcon} resizeMode='contain' />)}
+            text={feedbackSucceedText}
+            onDismiss={handleFeedbackModalDismiss}
+            action={(<NewHathorButton discrete title={t`See contract`} onPress={navigatesToNanoContractTransactions} />)}
+          />
```
Comment (pedroferreira1) [inline src/screens/NanoContract/NanoContractRegisterScreen.js:149]: Can the user click outside the modal and just close it without the `action` to navigate to the other screen?

File: src/screens/NanoContract/NanoContractRegisterScreen.js [hunk 27]
```diff
@@ -0,0 +1,323 @@
...
+    dispatch(nanoContractRegisterReady());
+  };
+
+  const navigatesToNanoContractTransactions = () => {
+    dispatch(nanoContractRegisterReady());
+    navigation.replace('NanoContractDetailsScreen', { ncId: formModel.ncId });
+  };
+
+  useEffect(() => {
+    dispatch(firstAddressRequest());
```
Comment (pedroferreira1) [inline src/screens/NanoContract/NanoContractRegisterScreen.js:131]: I don't think it makes sense to execute this request every time you register a nano contract. This will always be the same for the wallet, so you can check if it's already saved in redux to execute it only once.

File: src/screens/NanoContract/NanoContractRegisterScreen.js [hunk 28]
```diff
@@ -0,0 +1,323 @@
...
+  });
+
+  /* @param {
+   *   'nodeUrl'|'explorerUrl'|'explorerServiceUrl'|'walletServiceUrl'|'walletServiceWsUrl'
+   * } name */
+  const handleInputChange = useCallback(
+    (name) => (value) => {
+      if (isClean) {
+        setClean(false);
+      }
```
Comment (pedroferreira1) [inline src/screens/NanoContract/NanoContractRegisterScreen.js:87]: If `value === ''` shouldn't you set `isClean` to `true`?

File: src/sagas/nanoContract.js (hunk 22 above)
Comment (pedroferreira1) [reply src/sagas/nanoContract.js:120]: Yes, I know, but the method expects this parameter, we must send it. In the desktop wallet I used another request to get this information,

Comment (alexruzenhack) [reply src/sagas/nanoContract.js:120]: Added again.

**Review by pedroferreira1 — APPROVED** [review]

**Review by andreabadesso — APPROVED** [review]
